src/bundle.ts

```diff
@@ -0,0 +1,59 @@
+import { TranspiledFileWithSourceNode } from "./Transpile";
+import { SourceNode } from "source-map";
+
+const trimExtension = (fileName: string) => (fileName.endsWith(".ts") ? fileName.slice(0, -3) : fileName);
+const formatPath = (fileName: string) => trimExtension(fileName).replace("/", ".");
+
+export function bundleTranspiledFiles(
+    bundleFile: string,
+    entryModule: string,
+    transpiledFiles: TranspiledFileWithSourceNode[]
+): TranspiledFileWithSourceNode {
+    const moduleTable = moduleTableNode(transpiledFiles);
+    const requireOverride = `function require(file) return __Lua_BundleModules[file](); end\n`;
+    const entryPoint = `return require("${formatPath(entryModule)}")\n`;
+
+    const bundleNode = joinSourceChunks([moduleTable, requireOverride, entryPoint]);
+    const { code, map } = bundleNode.toStringWithSourceMap();
+
+    return {
+        fileName: bundleFile,
+        lua: code,
+        sourceMap: map.toString(),
+        sourceMapNode: moduleTable,
+        declaration: mergeDeclarations(transpiledFiles),
+        declarationMap: mergeDeclarations(transpiledFiles),
+    };
+}
+
+function moduleSourceNode(transpiledFile: TranspiledFileWithSourceNode): SourceNode {
+    const tableEntryHead = `["${formatPath(transpiledFile.fileName)}"] = function() `;
+    const tableEntryTail = `end,\n`;
+
+    if (transpiledFile.lua && transpiledFile.sourceMapNode) {
+        return joinSourceChunks([tableEntryHead, transpiledFile.sourceMapNode, tableEntryTail]);
+    } else {
+        return joinSourceChunks([tableEntryHead, tableEntryTail]);
+    }
+}
+
+function moduleTableNode(transpiledFiles: TranspiledFileWithSourceNode[]): SourceNode {
+    const tableHead = `__Lua_BundleModules = {\n`;
```

Comment:
`__Lua_BundleModules` looks quite inconsistent, can't we use names from https://github.com/TypeScriptToLua/TypeScriptToLua/pull/648/files#diff-7a7518949add62497c7fd54cdf9293e5?
 Also it probably should be local.

```suggestion
    const tableHead = `local ____modules = {\n`;
```